File: src/behavior.mjs
const ACTION_VERB =
  /\b(fix|add|implement|create|update|remove|refactor|debug|test|verify|run|deploy|migrate|search|find|explain|review)\b/i;
const CONSTRAINT =
  /\b(must|should|only|without|never|before|after|ensure|do not|don't|exactly|minimal)\b/i;
const ARTIFACT = /\b(file|function|class|module|api|endpoint|schema|migration|test|hook|rule)\b/i;
const VERIFY =
  /\b(tests?|verify|validate|lint|typecheck|build|rspec|jest|vitest|pytest|npm test|cargo test)\b/i;
const PRECISE_REJECT = /\b(wrong|incorrect|not working|still fails|try again|instead)\b/i;

function squash(rate, target) {
  if (target <= 0) return 0;
  return Math.max(0, Math.min(1, rate / target));
}

function avg(nums) {
  const v = nums.filter((n) => Number.isFinite(n));
  if (!v.length) return 0;
  return v.reduce((a, b) => a + b, 0) / v.length;
}

export function scoreBehaviorFromPrompts(prompts, meta = {}) {
  const real = (prompts || []).map((p) => String(p || "").trim()).filter(Boolean);
  const n = real.length;

  if (n === 0) {
    return {
      fluencyScore: 50,
      archetype: "Sprinter",
      realPromptCount: 0,
      dimensions: {
        briefing: 0.5,
        verification: 0.5,
        contextSetting: 0.5,
        iteration: 0.5,
        toolcraft: 0.5,
      },
      confidence: "low",
    };
  }

  let action = 0;
  let constraint = 0;
  let artifact = 0;
  let verify = 0;
  let precise = 0;
  let terse = 0;
  let delegate = 0;

  for (const text of real) {
    if (ACTION_VERB.test(text)) action++;
    if (CONSTRAINT.test(text)) constraint++;
    if (ARTIFACT.test(text)) artifact++;
    if (VERIFY.test(text)) verify++;
    if (PRECISE_REJECT.test(text)) precise++;
    if (text.length < 120) terse++;
    if (/\b(subagent|parallel|end-to-end|full pipeline|autonomous)\b/i.test(text)) delegate++;
    if (/^\/[a-z0-9-]+/i.test(text)) delegate++;
  }

  const briefing = squash(
    avg([action / n, constraint / n, artifact / n]),
    0.35
  );
  const verification = squash(verify / n, 0.15);
  const contextSetting = squash(artifact / n, 0.25);
  const iteration = squash(precise / n, 0.12);
  const toolcraft = squash(
    Math.min(1, (meta.toolCount || 0) / Math.max(1, n * 3)),
    0.6
  );

  const weights = {
    briefing: 0.24,
    verification: 0.22,
    contextSetting: 0.22,
    iteration: 0.18,
    toolcraft: 0.14,
  };

  const hedge = Math.min(1, n / 30);
  const raw =
    briefing * weights.briefing +
    verification * weights.verification +
    contextSetting * weights.contextSetting +
    iteration * weights.iteration +
    toolcraft * weights.toolcraft;

  const fluencyScore = Math.round((0.5 * (1 - hedge) + raw * hedge) * 100);

  const agencyVector = {
    delegate: delegate / n,
    terse: terse / n,
    briefing,
    verify: verification,
    precise: iteration,
  };

  let archetype = "Collaborator";
  const scores = {
    "Autonomous Agent": agencyVector.delegate * 2 + briefing,
    Architect: contextSetting * 2 + briefing,
    Debugger: precise * 2 + verification,
    Collaborator: 0.6,
    Sprinter: agencyVector.terse * 2 - verification * 0.5,
  };
  archetype = Object.entries(scores).sort((a, b) => b[1] - a[1])[0][0];

  return {
    fluencyScore,
    archetype,
    realPromptCount: n,
    dimensions: {
      briefing,
      verification,
      contextSetting,
      iteration,
      toolcraft,
    },
    confidence: n < 10 ? "low" : n < 40 ? "medium" : "high",
  };
}

File: analyzer/behavior.py
#!/usr/bin/env python3
"""Standalone optional behavior scorer (stdlib only). Not wired into the Node CLI."""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

ACTION = re.compile(
    r"\b(fix|add|implement|create|update|remove|refactor|debug|test|verify|run)\b",
    re.I,
)
CONSTRAINT = re.compile(r"\b(must|should|only|without|never|ensure)\b", re.I)


def _coerce_prompts(prompts) -> list[str]:
    if not isinstance(prompts, list):
        return []
    real: list[str] = []
    for p in prompts:
        if isinstance(p, str):
            text = p.strip()
        elif p is None:
            continue
        else:
            text = str(p).strip()
        if text:
            real.append(text)
    return real


def score_prompts(prompts: list[str]) -> dict:
    real = _coerce_prompts(prompts)
    n = len(real)
    if n == 0:
        return {"fluency_score": 50, "archetype": "Sprinter", "real_prompt_count": 0}

    action = sum(1 for p in real if ACTION.search(p))
    constraint = sum(1 for p in real if CONSTRAINT.search(p))
    briefing = min(1.0, (action / n + constraint / n) / 0.4)
    fluency = int(50 + briefing * 50)
    archetype = "Architect" if briefing > 0.6 else "Collaborator"
    return {
        "fluency_score": fluency,
        "archetype": archetype,
        "real_prompt_count": n,
        "briefing": round(briefing, 3),
    }


def main() -> int:
    if len(sys.argv) < 2:
        print("Usage: behavior.py <prompts.json>", file=sys.stderr)
        return 1
    path = Path(sys.argv[1])
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except OSError as err:
        print(f"error: cannot read {path}: {err}", file=sys.stderr)
        return 1
    except json.JSONDecodeError as err:
        print(f"error: invalid JSON in {path}: {err}", file=sys.stderr)
        return 1
    prompts = data if isinstance(data, list) else data.get("prompts", [])
    result = score_prompts(prompts if isinstance(prompts, list) else [])
    print(json.dumps(result, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
